### PR TITLE
Speed up results index

### PR DESCRIPTION
**Pull request by karyon**
By bringing it back to O(1) queries again

This speeds up the request time for students from 8s to 1.5s on my laptop and from 23 to 15s for staff users

Note that this is not the complete loading time for this page. my browser (firefox) needs another six seconds to finish processing/rendering the page.

### CODE REVIEW

File: evap/results/templates/results_index.html [hunk 1]
```diff
@@ -83,7 +83,8 @@ <h6 class="card-subtitle mb-1">{% trans 'Semester' %}</h6>
                     <tbody>
                     {% regroup evaluations by course as course_evaluations %}
                     {% for course, evaluations in course_evaluations|dictsort:"grouper.name" %}
-                        {% if course.evaluations.count > 1 %}
+                        {# we're actually using length here instead of count to avoid a query (`evaluations` is accessed below anyway #}
+                        {% if course.evaluations|length > 1 %}
```
Comment (karyon) [inline evap/results/templates/results_index.html]: this is broken, it should be `course.evaluations.all`. i'll probably prefetch the count instead.

File: evap/results/templates/results_index.html [hunk 2]
```diff
@@ -81,15 +81,15 @@ <h6 class="card-subtitle mb-1">{% trans 'Semester' %}</h6>
...
+                    {% for course, evaluations in courses %}
+                        {# this is the reason why results_index_evaluation does not need to include is_subentry in the cache key: an evaluation is a subentry or no subentry regardless of which user asks #}
+                        {% if course.num_evaluations > 1 %}
                             {% include 'results_index_course.html' %}
-                            {% for evaluation in evaluations|dictsort:"name" %}
+                            {% for evaluation in evaluations %}
                                 {% include 'results_index_evaluation.html' with links_to_results_page=evaluation|can_results_page_be_seen_by:request.user is_subentry=True %}
                             {% endfor %}
                         {% else %}
-                            {% for evaluation in evaluations|dictsort:"name" %}
```
Comment (karyon) [inline evap/results/templates/results_index.html]: @janno42 can you tell me why those sorts were here? perhaps to match what JS will sort by default so the user sees no change?

File: evap/results/tests/test_views.py [hunk 3]
```diff
@@ -14,13 +19,78 @@
...
 
-class TestResultsView(WebTestWith200Check):
+class TestResultsView(WebTest):
     url = '/results/'
-    test_users = ['manager']
 
-    @classmethod
-    def setUpTestData(cls):
-        mommy.make(UserProfile, username='manager', email="manager@institution.example.com")
+    @patch('evap.evaluation.models.Evaluation.can_be_seen_by', new=lambda self, user: True)
```
Comment (He3lixxx) [inline evap/results/tests/test_views.py]: I found that lambda part really hard to parse. The comma looks like it's separating the `new` argument from the next argument (which I thought was called `user`). Can we put parentheses around the `self, user` or do something else to improve this?

File: evap/results/views.py [hunk 4]
```diff
@@ -135,13 +136,26 @@ def index(request):
         additional_evaluations = get_evaluations_with_course_result_attributes(additional_evaluations)
         evaluations += additional_evaluations
 
-    evaluations.sort(key=lambda evaluation: (evaluation.course.semester.pk, evaluation.full_name))  # evaluations must be sorted for regrouping them in the template
+    # put evaluations into a dict that maps from course to a list of evaluations.
+    # this dict is sorted by course.pk (important for the zip below)
+    # (this relies on python 3.7's guarantee that the insertion order of the dict is preserved)
```
Comment (He3lixxx) [inline evap/results/views.py]: @janno42 do we have python3.7 running in production?

File: evap/results/views.py [hunk 5]
```diff
@@ -135,13 +136,26 @@ def index(request):
...
+    # annotate each course in courses with num_evaluations
+    annotated_courses = Course.objects.filter(pk__in=course_pks).annotate(num_evaluations=Count('evaluations')).order_by('pk').defer()
+    for course, annotated_course in zip(courses.keys(), annotated_courses):
+        course.num_evaluations = annotated_course.num_evaluations
+
+    degrees = Degree.objects.filter(courses__pk__in=course_pks).distinct()
+    course_types = CourseType.objects.filter(courses__pk__in=course_pks).distinct()
     template_data = dict(
-        evaluations=evaluations,
+        courses=courses.items(),
```
Comment (He3lixxx) [inline evap/results/views.py]: Waybe name the template variable `course_evaluation_tuples` or something? I was confused in the template that you didn't access it with `.items` and only saw later that you do it here.

A rather weak opinion of mine, though, so I'm also fine with leaving it like this.

File: evap/results/views.py [hunk 6]
```diff
@@ -135,13 +136,26 @@ def index(request):
...
+    # this dict is sorted by course.pk (important for the zip below)
+    # (this relies on python 3.7's guarantee that the insertion order of the dict is preserved)
+    evaluations.sort(key=lambda evaluation: evaluation.course.pk)
 
-    evaluation_pks = [evaluation.pk for evaluation in evaluations]
-    degrees = Degree.objects.filter(courses__evaluations__pk__in=evaluation_pks).distinct()
-    course_types = CourseType.objects.filter(courses__evaluations__pk__in=evaluation_pks).distinct()
+    courses = defaultdict(list)
+    for evaluation in evaluations:
+        courses[evaluation.course].append(evaluation)
```
Comment (He3lixxx) [inline evap/results/views.py]: Have you thought about doing something like ```
evaluation.course.num_evaluations = evaluation.course.num_evaluations + 1 if evaluation.course.num_evaluations else 0 ```
here instead of the query logic below?

File: evap/results/tests/test_views.py [hunk 7]
```diff
@@ -14,13 +19,78 @@
...
+        # then measure the number of queries with one more course and compare
+        make_course_with_evaluations('eggs')
+        call_command("refresh_results_cache")
+        with CaptureQueriesContext(connection) as context:
+            self.app.get(self.url, user="student")
+        num_queries_after = context.final_queries - context.initial_queries
+
+        self.assertEqual(num_queries_before, num_queries_after)
+
+        # django does not clear the LocMemCache in between tests. clear it hear just to be safe.
```
Comment (karyon) [inline evap/results/tests/test_views.py]: typo

File: evap/results/views.py (hunk 4 above)
Comment (karyon) [reply evap/results/views.py]: yes you do :)

File: evap/results/views.py (hunk 6 above)
Comment (karyon) [reply evap/results/views.py]: I don't quite understand, where does `num_evaluations` come from if not from the query logic?

File: evap/results/views.py (hunk 5 above)
Comment (karyon) [reply evap/results/views.py]: I called this courses_and_evaluations, but it's still the same in template and view. actually I wanted to do the `.items()` in the template but that does not work, seems like i still haven't understood the rules how method calls work there...

Comment (He3lixxx) [reply evap/results/views.py]: Hmm, strange. A bit of debugging indicates that using `.items` on a `defaultdict` in a template breaks (returns an empty list), if you convert it to a normal dict first, it works fine.

I guess I'm fine with the current code.

File: evap/results/views.py (hunk 6 above)
Comment (He3lixxx) [reply evap/results/views.py]: We can simply count up the count of evaluations from 0 in this loop, it runs once for every evaluation anyway.

Doing it that way instead of querying the database for the number of evaluations would save that query -  but I'm not sure about whether it's better.

Comment (karyon) [reply evap/results/views.py]: `evaluations` does not contain all evaluations, but is filtered. here i actually need all evaluations. should i add a comment in that regard?

Comment (He3lixxx) [reply evap/results/views.py]: Ah, I see. No, I think it's fine, I should have just read more context.

**Comment by He3lixxx** [comment]
🎉

**Comment by He3lixxx** [comment]
You removed the `stdout=StringIO()` parameter from the `management.call_command()` call - if was like this before.
```
management.call_command('refresh_results_cache', stdout=StringIO())
```
Did you _want_ to do this? I think I tested it and everything was fine - but with the current master, the stdout output is printed to the console again.

**Comment by karyon** [comment]
No, that was for debugging, i guess. I opened https://github.com/fsr-de/EvaP/pull/1399.

**Comment by karyon** [comment]
thanks for spotting!
